psh/du/du.c

```diff
@@ -0,0 +1,317 @@
+ */
+static int isDone(dev_t dev, ino_t inum, nlink_t nlink)
+{
+	struct already *ap;
+	struct already **pap = &du_common.already[(int)inum % ALREADY_MAX];
+
+	for (;;) {
+		ap = *pap;
+		if (ap == NULL) {
+			break;
+		}
+
+		if ((ap->inum == inum) && (ap->dev == dev)) {
+			ap->nlink--;
+			if (ap->nlink == (nlink_t)0) {
+				*pap = ap->next;
+				free(ap);
+			}
+			return 1;
+		}
+
+		pap = &ap->next;
+	}
+
+	ap = (struct already *)malloc(sizeof(*ap));
+	if (ap == NULL) {
+		fprintf(stderr, "du: out of memory\n");
+		return -1;
+	}
+
+	ap->next = NULL;
+	ap->inum = inum;
+	ap->dev = dev;
+	ap->nlink = nlink - 1;
+	*pap = ap;
+
+	return 0;
+}
+
+
+static void freeAlreadyList(void)
+{
+	for (int i = 0; i < ALREADY_MAX; ++i) {
+		struct already *list = du_common.already[i];
+		while (list != NULL) {
+			struct already *temp = list;
+			list = list->next;
+			free(temp);
+		}
+	}
+}
+
+
+static size_t roundPrefix(size_t v)
+{
+	switch (du_common.ki) {
+		case 1: return (v + (size_t)999) / (size_t)1000;  /* kilo */
+		case 2: return (v + (size_t)1023) / (size_t)1024; /* Kibi */
+		default: break;
+	}
+	return v;
+}
+
+
+/*
+ * doDirectory - process the directory.
+ * Return the size (in bytes) of the directory and its descendants.
+ */
+static size_t doDirectory(size_t startPos, int curLevel, dev_t dev)
+{
+	DIR *dp;
+	bool maybePrint;
+	struct stat stbuf;
+	size_t total = 0;
+	struct dirent *entry;
+	char *d = du_common.dent;
+
+	if (lstat(d, &stbuf) < 0) {
+		fprintf(stderr, "du: %s: %s\n", d, strerror(errno));
+		return (size_t)0;
+	}
+
+	if ((stbuf.st_dev != dev) && (dev != 0) && (du_common.crosschk)) {
+		return (size_t)0;
+	}
+
+	if (stbuf.st_size > (off_t)0) {
+		total = (size_t)stbuf.st_size;
+	}
+
+	switch (stbuf.st_mode & S_IFMT) {
+		case S_IFDIR:
+			/*
+			 * Directories should not be linked except to "." and "..", so this
+			 * directory should not already have been done.
+			 */
+			maybePrint = !du_common.silent;
+			dp = opendir(d);
+			if (dp == NULL) {
+				break;
+			}
+
+			for (;;) {
+				du_common.dent[startPos] = '\0';
+				entry = readdir(dp);
+				if (entry == NULL) {
+					break;
+				}
+
+				if (!((entry->d_name[0] == '.') && ((entry->d_name[1] == '\0') || ((entry->d_name[1] == '.') && (entry->d_name[2] == '\0'))))) {
+					const size_t nextPos = makePath(startPos, entry->d_name);
+					if (nextPos > 0u) {
+						total += doDirectory(nextPos, curLevel - 1, stbuf.st_dev);
+					}
+				}
+			}
+
+			closedir(dp);
+			break;
+
+		case S_IFBLK:
+		case S_IFCHR:
+			/* st_size for special files is not related to blocks used. */
+			total = (size_t)0;
+			/* fall-through */
+
+		default:
+			if ((stbuf.st_nlink > (nlink_t)1) && (isDone(stbuf.st_dev, stbuf.st_ino, stbuf.st_nlink) != 0)) {
+				return (size_t)0;
+			}
+			maybePrint = du_common.all;
+			break;
+	}
+
+
+	if ((curLevel >= du_common.levels) || ((maybePrint) && (curLevel >= 0))) {
+		printf("%zu\t%s\n", roundPrefix(total), d);
```

Comment:
Perhaps this can be used: https://github.com/phoenix-rtos/phoenix-rtos-utils/blob/7b6b834b43cc1a5df57c10b5f9756cb973a9ba25/psh/pshapp/pshapp.c#L210

To not reimplement existing functionality